### PR TITLE
Simplify HiveDataSource::addSplit

### PR DESCRIPTION
**Pull request by yingsu00**

### CODE REVIEW

File: velox/connectors/hive/HiveDataSource.cpp [hunk 1]
```diff
@@ -585,20 +585,7 @@ void HiveDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
...
-          scanChildSpec, VARCHAR(), velox::variant(split_->filePath));
-    } else if (fieldName == kBucket) {
-      if (split_->tableBucketNumber.has_value()) {
-        setConstantValue(
-            scanChildSpec,
-            INTEGER(),
-            velox::variant(split_->tableBucketNumber.value()));
-      }
-    } else if (!fileType->containsChild(fieldName)) {
+    if (!fileType->containsChild(fieldName)) {
```
Comment (Yuhta) [inline velox/connectors/hive/HiveDataSource.cpp:588]: Just merge the 3 in an empty branch and do nothing in these cases?  this way we can avoid allocating null constants then overwrite them later.  Or comment that they will be overwritten later here.

Comment (yingsu00) [reply velox/connectors/hive/HiveDataSource.cpp:588]: Thanks for brainstorming. Currently, the loop is on every output column (readerOutputType_), but the columns with filters may not be output columns. Also, scanSpec_->childByName(entry.first); is an O(n) operation. How about looping over scanSpec_ children which contains both? This way we will have the loop once, and we will be able to remove the scanSpec_->childByName calls. I'll try it and update the PR tomorrow.

**Comment by yingsu00** [comment]
@Yuhta I have updated the PR, will you please review it again? Thanks!

File: velox/connectors/hive/HiveDataSource.cpp [hunk 2]
```diff
@@ -579,62 +579,49 @@ void HiveDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
   }
 
   auto& fileType = reader_->rowType();
+  std::vector<TypePtr> fileColumnTypes = fileType->children();
 
-  std::vector<TypePtr> columnTypes = fileType->children();
-  for (int i = 0; i < readerOutputType_->size(); i++) {
-    auto fieldName = readerOutputType_->nameOf(i);
-    auto scanChildSpec = scanSpec_->childByName(fieldName);
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSource.cpp:586]: A question: does the old code need to do nullable check here? Or we assume scan spec has one spec for each reader output child type? Thanks!

File: velox/connectors/hive/HiveDataSource.cpp [hunk 3]
```diff
@@ -579,62 +579,49 @@ void HiveDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
...
+  auto childrenSpecs = scanSpec_->children();
+  for (int i = 0; i < childrenSpecs.size(); i++) {
+    auto childSpec = childrenSpecs[i].get();
+    std::string fieldName = childSpec->fieldName();
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSource.cpp:587]:
```
const std::string fieldName
```

File: velox/connectors/hive/HiveDataSource.cpp [hunk 4]
```diff
@@ -579,62 +579,49 @@ void HiveDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
...
-    }
-  }
-
-  // Set constant values for partition keys and $path column. If these are
-  // used in filters only, the loop above will miss them.
-  for (const auto& entry : split_->partitionKeys) {
-    auto childSpec = scanSpec_->childByName(entry.first);
-    if (childSpec) {
-      setPartitionValue(childSpec, entry.first, entry.second);
+      auto outputTypeIdx = readerOutputType_->getChildIdxIfExists(fieldName);
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSource.cpp:602]: in which case fieldName doesn't exist in readerOutputType_? Is it a valid case or shall we add a check here? thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@yingsu00 thanks for the cleanup. The code simpler % some questions.

File: velox/connectors/hive/HiveDataSource.cpp (hunk 2 above)
Comment (yingsu00) [reply velox/connectors/hive/HiveDataSource.cpp:586]: > A question: does the old code need to do nullable check here? Or we assume scan spec has one spec for each reader output child type? Thanks!

Yes scan spec has one spec for each reader output child type. See HiveDataSource::makeScanSpec()
```
// Process columns that will be projected out.
  for (int i = 0; i < rowType->size(); ++i) {
    auto& name = rowType->nameOf(i);
    auto& type = rowType->childAt(i);
    auto it = outputSubfields.find(name);
    if (it == outputSubfields.end()) {
      spec->addFieldRecursively(name, *type, i);
      filterSubfields.erase(name);
      continue;
    }
```

File: velox/connectors/hive/HiveDataSource.cpp (hunk 4 above)
Comment (yingsu00) [reply velox/connectors/hive/HiveDataSource.cpp:602]: > in which case fieldName doesn't exist in readerOutputType_? Is it a valid case or shall we add a check here? thanks!

For example, some fields/columns are used in the filter but not projected out. For such cases, these fields/columns need to be read to evaluate filters for not one of the readerOutputType_.

**Review by xiaoxmeng — APPROVED** [review]
@yingsu00 LGTM. Can you fix the circle ci failure? thanks!

File: velox/connectors/hive/HiveDataSource.h [hunk 5]
```diff
@@ -107,7 +107,7 @@ class HiveDataSource : public DataSource {
   dwio::common::ReaderOptions readerOpts_;
   memory::MemoryPool* pool_;
   VectorPtr output_;
-  RowTypePtr readerOutputType_;
+  RowTypePtr outputType_;
```
Comment (Yuhta) [inline velox/connectors/hive/HiveDataSource.h:110]: `outputType_` is not a proper name, `readerOutputType_` is data source / table scan output + remaining filter inputs, I would not change this name

**Review by Yuhta — COMMENTED** [review]
Other than the renaming of `readerOutputType_` everything is good.  Although we don't really need to set constant values for filter only specs.

**Comment by yingsu00** [comment]
@Yuhta @xiaoxmeng Thank you very much, I have updated the PR with readerOutputType_ named back. 

> Although we don't really need to set constant values for filter only specs.

Could you explain a bit about this? In this change, the filter only specs was not set constant. See this code
```
auto outputTypeIdx = readerOutputType_->getChildIdxIfExists(fieldName);
      if (outputTypeIdx.has_value()) {. 
        // childSpec is for a column that is an output column
        auto outputIdx = outputTypeIdx.value();
        auto fileTypeIdx = fileType->getChildIdxIfExists(fieldName);
        if (!fileTypeIdx.has_value()) {
          // Column is missing. Most likely due to schema evolution.
          setNullConstantValue(
              childSpec, readerOutputType_->childAt(outputIdx));
        } else {.
          // We know the fieldName exists in the file, make the type at that
          // position match what we expect in the output.
          fileColumnTypes[fileTypeIdx.value()] =
              readerOutputType_->childAt(outputIdx);
          childSpec->setConstantValue(nullptr);
        }
      }
// No else here, and there is no block to handle the case where childSpec is not an output column.
```

**Review by Yuhta — APPROVED** [review]

**Comment by Yuhta** [comment]
You set the partition key etc. according to spec instead of output type, so filter only fields are also set.  It is correct but not necessary because filter only constant fields are only examined for split skipping purpose and not checked for each row.

**Pull request closed without merging**